docs/api.rst

```diff
@@ -104,3 +105,98 @@ The ``pika/execution.hpp`` header provides functionality related to ``std::execu
 .. literalinclude:: ../examples/documentation/when_all_vector_documentation.cpp
    :language: c++
    :start-at: #include
+
```

Comment:
The API reference page is starting to get quite long. We could consider splitting it up to have one header per page. I'm still slightly in favour of keeping it on one page for Ctrl-F-ability and clicking on references not changing the page, but wouldn't object to splitting it up either.